File: Stacks/NextGreaterEle2/nextGreat.cpp
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        if(n==1)return {-1};
        vector<int>nge(n,-1);
        stack<int>st;
        for(int i=n-1;i>=0;i--){
            st.push(nums[i]);
        }
        for(int i=n-1;i>=0;i--){
            while(st.size()!=0 && nums[i]>=st.top())st.pop();
            if(st.size())nge[i]=st.top();
            st.push(nums[i]);
        }
        return nge;
    }
};

File: Stacks/NextGreaterEle2/nextGreat.py
from typing import List

class Solution:
    def nextGreaterElements(self, arr: List[int]) -> List[int]:
        actualLen = len(arr)
        arr = arr+arr[0:actualLen-1]
        n = len(arr)
        st = []
        ans = [-1] * n
        st.append(arr[n-1])
        for i in range(n-2,-1,-1):
            while len(st) and arr[i]>=st[-1]: st.pop()
            if len(st): ans[i]=st[-1]
            st.append(arr[i])
        return ans[0:actualLen]